### DISCUSSION TITLE
SignedProposal Has No Signature?

### DISCUSSION THREAD

**Discussion by jonmchan** [hyperledger/fabric-gateway#357] [Q&A] [answered]
I noticed that none of the implementations utilize the signature field in the SignedProposals in the proposal_builder (for example - https://github.com/hyperledger/fabric-gateway/blob/main/pkg/client/proposalbuilder.go#L42). Could anyone explain the reasoning for this? Without that, can't any parameter be changed in the proposal and be accepted as valid? I don't see any other place where the content is being signed or hashed.

**Comment by jonmchan** [comment]
It dawned on me - it isn't automatically signed to support offline signing. I now see there is a `setSignature` method in the proposal. By not automatically signing it, we can send it to be offline signed in the event that the private key is not in the identity. Would this be accurate?

**Comment by bestbeforetoday** [comment, marked as answer]
The proposal is always signed just before it is sent using the signer specified when connecting the Gateway, unless it has already been signed using the off-line signing flow:

- Endorse: https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/pkg/client/proposal.go#L64
- Evaluate: https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/pkg/client/proposal.go#L104

The off-line signing flow sets the supplied signature, which makes the signing during the endorse and evaluate a no-op since the signature is already present:

- https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/pkg/client/gateway.go#L194

The reason a signature isn’t created when the proposal is created is that, if you are exclusively using off-line signing, you can connect the Gateway without a signer. In this case, creating the signature using the Gateway’s signer when the proposal is created would fail.

**Reply by jonmchan** [reply]
I'll study the offline scenario tests. I appreciate the testing on this project! Thanks!
